### PR TITLE
Feat: Streaming frontend and backend (WIP)

### PR DESCRIPTION
**Pull request by tobitege**
**What is the problem that this fixes or functionality that this introduces? Does it fix any open issues?**

- Overall goal: streaming for frontend and backend.
- Streaming idea picked from other PR, it doesn't stream LLM, though.
- First item in this PR was to get the terminal hooked up for streaming in the frontend.
- Revised ssh_box.py:
  - incl. `[PEXPECT]$` output handling
  - SSHExecCancellableStream is now threaded and has an output queue.
  - Added a bash_command_splitter.py utility incl. unit test (refactor of split_bash_commands in ssh_box)
  - DockerSSHBox.execute(): added `[PEXPECT]$` output handling
- Touched probably a couple too many files, that are now gone (probably need to rebase or something).
- The mixin.py streams feedback as it is generated (in chunks of 256 bytes) from the sandbox to main terminal.
    - "_source_bashrc": split up into 2 separate commands to avoid stream locking (might be temporary)
    - "_handle_stream_output" is doing a lot of work.
    - need to check what happened to plugin_initialized in the meantime
    - PluginMixin() slightly changed logging order
Other changes in interfaces to make avoid errors, I don't recall them all right now.

**Open issues:**
The overall integration is open, i.e. the major problem I experienced was that the LLM traffic was out of sync and caused errors, but I hadn't much time since then to look into it deeper. If I disable streaming the problems were gone.
Running into blockers in sandbox-related tests.

**Give a brief summary of what the PR does, explaining any non-trivial design decisions**

Stream output of commands run inside sandbox back as it happens, not only when finished.
Threaded, with timeouts, extended stream end detections.

**Other references**

https://github.com/OpenDevin/OpenDevin/issues/1193
https://github.com/OpenDevin/OpenDevin/issues/2754

### CODE REVIEW

**Comment by tobitege** [comment]
Need to fix errors after rebase now.

**Review by xingyaoww — COMMENTED** [review]
Thanks for taking this! i did a quick look, though i didn't really go into details.
Looks reasonable on the skim, but i do have a high-level question, how do you plan to handle the interaction between BE and FE to streaming using event stream without polluting the whole stream (since the agent also uses it to get histories)?

Are you thinking about streaming token-by-token in the event stream (each time a JSON with *some* tokens, hide this from the agent), and then send one final complete output as a confirmation at the end of streaming, which the agent can see?

**Comment by tobitege** [comment]
> Thanks for taking this! i did a quick look, though i didn't really go into details. Looks reasonable on the skim, but i do have a high-level question, how do you plan to handle the interaction between BE and FE to streaming using event stream without polluting the whole stream (since the agent also uses it to get histories)?
> 
> Are you thinking about streaming token-by-token in the event stream (each time a JSON with _some_ tokens, hide this from the agent), and then send one final complete JSON (over event stream) as a confirmation at the end of streaming, which the agent can see?

Ideally, yes. The frontend streaming right now is for the terminal. And for that I'd like to use the EventStream to capture any responses coming in from bash commands real-time, not just when finished.
The problem will then indeed be to only send a "full" (collected) reply as the observation(?) into the "regular" stream.
If that is not feasible with the current event stream, maybe we need a 2nd one for general UI streaming purposes (if that is feasible)?
But I'm not that far yet, working on the runtime/server runtime to work with the changes.

File: opendevin/runtime/utils/bash_command_splitter.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
+import re
+
+
+def split_bash_commands(command_string):
```
Comment (yufansong) [inline opendevin/runtime/utils/bash_command_splitter.py:4]: I think `split_bash_commands` and some following test are target for SSHBOX. I think maybe we will deprecated the SSHBOX in the future. Not sure whether it worth your effort on it.

**Review by yufansong — COMMENTED** [review]
I worried some of change in this PR will be depcreated after we use `od-runtime-client`.

File: opendevin/runtime/plugins/mixin.py
[the remark this answers is not in the document]
Comment (tobitege) [reply opendevin/runtime/plugins/mixin.py:1]: That's fine. Maybe your runtime client makes this PR obsolete by default.

File: opendevin/runtime/utils/bash_command_splitter.py (hunk 1 above)
Comment (tobitege) [reply opendevin/runtime/utils/bash_command_splitter.py:4]: There is at least a new unit test for it.

**Comment by tobitege** [comment]
Actually closing this in favor of #2602 where more work has been done with regard to async support.
This PR may still contain some nuggest, like the revised split_bash_commands and its unit test.

**Pull request closed without merging**